### PR TITLE
[Prometheus Receiver] Adding testcases for core metrics

### PR DESCRIPTION
**Pull request by PaurushGarg**
_Note: This PR is not dependent on previous PRs, and all the change in this PR belongs to this PR._
**Description:** 
- Modified testcases for Prometheus core metrics to cover the `start_timestamp` issues. 
- Modified `isFirstFailedScrape` method to validate first failed scrape on the basis of flags provided by OTLP format for staleness markers. For now skipped this validation with comment, as current Prometheus Receiver metric builder is not directly using OTLP format.
- Modified `setupMockPrometheus` method to separate mockPrometheus response and preparation of receiver configs. This is required to keep modification of config by different testcases  at one place.
- Modified `testComponent` method to accept custom receiver configs. This is required to avoid multiple copies of this function (because of custom configs) with every testcase.
- Modified external_labels test to remove `testComponent` copy and use method  `testComponent` in metric_receiver_helper_test.go 

**Link to tracking Issue:** <Issue number if applicable>
Issue Number: 6000
cc @alolita @Aneurysm9

### CODE REVIEW

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 1]
```diff
@@ -207,14 +242,57 @@ func getValidScrapes(t *testing.T, rms []*pdata.ResourceMetrics) []*pdata.Resour
...
+			}
+		case pdata.MetricDataTypeSummary:
+			for i := 0; i < m.Summary().DataPoints().Len(); i++ {
+				stalenessFlag = m.Summary().DataPoints().At(i).Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue)
+				if !stalenessFlag {
+					return false
+				}
+			}
+		}
+	}
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_helper_test.go:295]: I don't think `stalenessFlag` is needed since the function will return as soon as any non-stale metric is found.

```suggestion
	for _, m := range metrics {
		switch m.Name() {
		case "up", "scrape_duration_seconds", "scrape_samples_scraped", "scrape_samples_post_metric_relabeling", "scrape_series_added":
			continue
		}
		switch m.DataType() {
		case pdata.MetricDataTypeGauge:
			for i := 0; i < m.Gauge().DataPoints().Len(); i++ {
				if !m.Gauge().DataPoints().At(i).Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
					return false
				}
			}
		case pdata.MetricDataTypeSum:
			for i := 0; i < m.Sum().DataPoints().Len(); i++ {
				if !m.Sum().DataPoints().At(i).Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
					return false
				}
			}
		case pdata.MetricDataTypeHistogram:
			for i := 0; i < m.Histogram().DataPoints().Len(); i++ {
				if !m.Histogram().DataPoints().At(i).Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
					return false
				}
			}
		case pdata.MetricDataTypeSummary:
			for i := 0; i < m.Summary().DataPoints().Len(); i++ {
				if !m.Summary().DataPoints().At(i).Flags().HasFlag(pdata.MetricDataPointFlagNoRecordedValue) {
					return false
				}
			}
		}
	}
```

File: receiver/prometheusreceiver/metrics_receiver_test.go [hunk 2]
```diff
@@ -28,6 +28,10 @@ import (
 
 // target1 has one gauge, two counts of a same family, one histogram and one summary. We are expecting the both
 // successful scrapes will produce all metrics using the first scrape's timestamp as start time.
+
+// With the 2nd page, we are simulating a reset (values smaller than previous), start_timestamp should be from
+// this run for the 2nd scrape, despite a failed scrape between 1st and 2nd page.
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_test.go:33]: Is this change necessary?  Isn't there a separate test that tests this scenario?  If not, what of the situation that was previously tested, i.e., a successful scrape after a failed scrape that is not a series reset?

File: receiver/prometheusreceiver/metrics_receiver_test.go [hunk 3]
```diff
@@ -589,15 +597,49 @@ rpc_duration_seconds_sum{foo="no_quantile"} 101
 rpc_duration_seconds_count{foo="no_quantile"} 55
 `
 
+var target3Page3 = `
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_test.go:600]: What is this testing?  Please add some comments explaining its utility.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go (hunk 1 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_helper_test.go]: @Aneurysm9 thanks. I have removed the stalenessFlag accordingly.

File: receiver/prometheusreceiver/metrics_receiver_test.go (hunk 2 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_test.go:33]: The scenario: _highervalue-failescrape-lowervalue_ was not tested before for summary and histogram. Receiver is not resetting the `start_timestamp` for all cumulative types as per this scenario, and this was not captured in these tests. 

Earlier this test case tested: _value-failedscrape-highervalue_ I have added this scenario in target3page3. Target3 checks the histogram and summary only. So I added this scenario there. 

For counter, both the scenarios- _highvalue-failedscrape-lowervalue_; _value-failedscrape-highervalue_ is tested in target2.

File: receiver/prometheusreceiver/metrics_receiver_test.go (hunk 3 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_test.go:603]: @Aneurysm9 thanks. This is added as I changed the scenario of target1 to reflect _high value-failedscrape-lowervalue_. The scenario I removed from target1, I added in target3 using target3page3. 
I have provided the comments there now.

Comment (Aneurysm9) [reply receiver/prometheusreceiver/metrics_receiver_test.go:603]: I would do that the other way around, then.  Don't change target1 and add the new scenario to target3.  That way only one scenario changes and it is less likely to cause confusion.

Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_test.go]: @Aneurysm9 thanks. I have now made changes accordingly.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 4]
```diff
@@ -140,22 +165,38 @@ func setupMockPrometheus(tds ...*testData) (*mockPrometheus, *promcfg.Config, er
...
+			for _, scrapeConfig := range pCfg.ScrapeConfigs {
+				scrapeConfig.LabelLimit = promConfig.labelLimit
+			}
+		}
+		if promConfig.renamingCfg != nil {
+			for _, scrapeConfig := range pCfg.ScrapeConfigs {
+				scrapeConfig.MetricRelabelConfigs = promConfig.renamingCfg
+			}
+		}
+	}
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_helper_test.go:197]: This is going to be cumbersome and error-prone to maintain.  I'd prefer to use a mutator that can be passed to `testComponent()`.  See how I did it in [this PR](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6421/files#diff-ad6f97323c1aea65ad2748924e71b6f180a249d67a68151534874f1b0965d1e9R502), which might remove the need for these changes anyway when it lands.

File: receiver/prometheusreceiver/metrics_receiver_test.go [hunk 5]
```diff
@@ -578,26 +593,24 @@ corrupted_hist_count 15
...
-	require.Greater(t, len(resourceMetrics), 0, "At least one resource metric should be present")
-	m1 := resourceMetrics[0]
-	// m1 has 3 metrics + 5 internal scraper metrics
-	assert.Equal(t, 8, metricsCount(m1))
-
 	wantAttributes := td.attributes
 
+	m1 := resourceMetrics[0]
+	// m1 has 3 metrics + 5 internal scraper metrics
+	assert.Equal(t, 9, metricsCount(m1))
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_test.go:613]: 3+5 != 9

File: receiver/prometheusreceiver/metrics_receiver_test.go [hunk 6]
```diff
@@ -648,8 +682,7 @@ func verifyTarget3(t *testing.T, td *testData, resourceMetrics []*pdata.Resource
 
 	m2 := resourceMetrics[1]
 	// m2 has 3 metrics + 5 internal scraper metrics
-	assert.Equal(t, 8, metricsCount(m2))
-
+	assert.Equal(t, 9, metricsCount(m2))
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_test.go:685]: 3+5 != 9

**Review by Aneurysm9 — CHANGES_REQUESTED** [review]
Please unwind the changes to `testComponent()` and related functions.  We should prefer the config mutator function approach instead.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go (hunk 4 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_helper_test.go]: @Aneurysm9 thanks. I have now updated accordingly.

File: receiver/prometheusreceiver/metrics_receiver_test.go (hunk 6 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_test.go]: @Aneurysm9 thanks. I have now updated.

File: receiver/prometheusreceiver/metrics_receiver_test.go (hunk 5 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_test.go]: @Aneurysm9 thanks. I have now updated.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 7]
```diff
@@ -455,90 +492,54 @@ func compareSummary(count uint64, sum float64, quantiles [][]float64) summaryPoi
 		req := assert.Equal(t, len(quantiles), summaryDataPoint.QuantileValues().Len())
 		if req {
 			for i := 0; i < summaryDataPoint.QuantileValues().Len(); i++ {
-				assert.Equal(t, quantiles[i][0], summaryDataPoint.QuantileValues().At(i).Quantile(), "Summary quantile do not match")
-				assert.Equal(t, quantiles[i][1], summaryDataPoint.QuantileValues().At(i).Value(), "Summary quantile values do not match")
+				assert.Equal(t, quantiles[i][0], summaryDataPoint.QuantileValues().At(i).Quantile(),
+					"Summary quantile do not match")
+				if math.Float64bits(quantiles[i][1]) == value.NormalNaN {
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_helper_test.go:497]: I thought NaN values were [not comparable](https://go.dev/play/p/pw6aEeYklDk).  You can use `math.IsNaN()` or `math.Float64bits()`, which is the way `value.IsStaleNaN()` works:

```go
math.Float64bits(v) == value.NormalNaN
```

That should only be required if it is necessary to distinguish between the normal and stale NaN values, otherwise `math.IsNaN()` should suffice.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 8]
```diff
@@ -110,11 +112,11 @@ var (
...
 type testData struct {
-	name           string
-	pages          []mockPrometheusResponse
-	attributes     pdata.AttributeMap
-	useOpenMetrics bool
-	validateFunc   func(t *testing.T, td *testData, result []*pdata.ResourceMetrics)
+	name             string
+	pages            []mockPrometheusResponse
+	attributes       pdata.AttributeMap
+	skipValidScrapes bool
```
Comment (Aneurysm9) [inline receiver/prometheusreceiver/metrics_receiver_helper_test.go:118]:
```suggestion
	validateScrapes bool
```

This should probably be phrased in the affirmative and better describe what it affects.  I initially thought that this would cause the tests to skip testing of valid scrapes and only test invalid scrapes, but instead it is controlling whether the function called `getValidScrapes()` is invoked.  Another way to phrase it might be `skipScrapeValidation`.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 9]
```diff
@@ -355,11 +392,11 @@ func compareAttributes(attributes map[string]string) numberPointComparator {
...
 		if req {
 			for k, v := range attributes {
-				value, ok := numberDataPoint.Attributes().Get(k)
+				val, ok := numberDataPoint.Attributes().Get(k)
 				if ok {
-					assert.Equal(t, v, value.AsString(), "Attributes do not match")
+					assert.Equal(t, v, val.AsString(), "Attributes do not match")
 				} else {
-					assert.Fail(t, "Attributes key do not match")
+					assert.Failf(t, "Attributes key do not match", k)
```
[the remark this answers is not in the document]
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_helper_test.go]: @Aneurysm9 thanks. I have updated.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go [hunk 10]
```diff
@@ -371,11 +408,11 @@ func compareSummaryAttributes(attributes map[string]string) summaryPointComparat
...
 		if req {
 			for k, v := range attributes {
-				value, ok := summaryDataPoint.Attributes().Get(k)
+				val, ok := summaryDataPoint.Attributes().Get(k)
 				if ok {
-					assert.Equal(t, v, value.AsString(), "Summary attributes value do not match")
+					assert.Equal(t, v, val.AsString(), "Summary attributes value do not match")
 				} else {
-					assert.Fail(t, "Summary attributes key do not match")
+					assert.Failf(t, "Summary attributes key do not match", k)
```
[the remark this answers is not in the document]
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_helper_test.go]: @Aneurysm9 thanks. I have updated.

File: receiver/prometheusreceiver/metrics_receiver_helper_test.go (hunk 7 above)
Comment (PaurushGarg) [reply receiver/prometheusreceiver/metrics_receiver_helper_test.go]: @Aneurysm9 thanks. I have updated.

File: exporter/prometheusexporter/end_to_end_test.go [hunk 11]
```diff
@@ -41,7 +41,7 @@ func TestEndToEndSummarySupport(t *testing.T) {
 	}
 
 	// 1. Create the Prometheus scrape endpoint.
-	waitForScrape := make(chan bool, 1)
+	waitForScrape := make(chan bool, 10)
```
Comment (Aneurysm9) [inline exporter/prometheusexporter/end_to_end_test.go:44]: This change doesn't belong here, does it?

Comment (PaurushGarg) [reply exporter/prometheusexporter/end_to_end_test.go]: @Aneurysm9 oh no! sorry. I left it there by mistake. Didn't notice also. reverted it back to its original.

**Review by Aneurysm9 — APPROVED** [review]
Please squash this to a single commit before opening upstream to minimize the noise regarding prior approaches that ends up in the commit message when it is merged.

**Comment by PaurushGarg** [comment]
Merged
